code-dot-org/code-dot-org
# add dismissible notification when student scores are available

Finishes implementation work for [AITT-816: Alert teachers of waiting scores - US 3](https://codedotorg.atlassian.net/browse/AITT-816). all that remains is some fit-and-finish changes and to launch the experiment.

the requirement is to show the alert only once on each lesson. to implement this, I'm storing a map of lesson positions for each user. this makes it so that dismissing the alert in one lesson (e.g. lesson 11) in one of the many copies of csd3 (e.g. `interative-animations-games-2024`) will also hide the alert on the same lesson in other copies of the same unit (e.g. `csd3-2024`). 

The alerts are limited to csd-2023 and csd-2024 so that we do not unintentionally keep showing these alerts beyond the current version year -- if we decide to show these for 2025 that will be easy to change later.

## screenshots

each sequence starts with clearing the user's settings so that they have not "seen" the alert on any lessons.

### dismiss alert via close button

https://github.com/user-attachments/assets/7845ce07-e4e1-4a04-b86b-16b6af323b40

### view scores

https://github.com/user-attachments/assets/7adde5c8-bc22-4dbb-836a-eebb14969a3d

### clicking fab dismisses alert

https://github.com/user-attachments/assets/a9b83782-0dd9-4e01-b171-aeb826e8cf87

### auto dismiss alert if TA window is already open

https://github.com/user-attachments/assets/e47906e5-2bf6-4223-a072-f9484b9b0def

## Testing story
* manual testing as shown in screenshots
* js tests
* ruby integration tests instead of controller tests, which make it possible to test the interaction between routes from different controllers
* no green check for drone yet, but I do now have a passing [unit](https://drone.cdn-code.org/code-dot-org/code-dot-org/18997/1/6) and [UI test](https://drone.cdn-code.org/code-dot-org/code-dot-org/18961/2/6) run across multiple drone runs
